### PR TITLE
Adding support for blocking service/tags that wont work with the dns interface.

### PR DESCRIPTION
**Pull request by vamage**
Creating new config option EnforceValidDNS to force the validation of the registration of new services. 
Its currently disabled by default. 

#3202

### CODE REVIEW

File: agent/consul/catalog_endpoint.go [hunk 1]
```diff
@@ -97,6 +99,22 @@ func (c *Catalog) Register(args *structs.RegisterRequest, reply *struct{}) error
 		}
 	}
 
+	if c.srv.config.EnforceValidDNS {
+		dnsNameRe := regexp.MustCompile(`^[a-zA-Z0-9_-]+$`)
```
Comment (preetapan) [inline agent/consul/catalog_endpoint.go]: This incurs the cost of recompiling the same regex over and over for every register call, instead dnsNameRegex should be compiled once into an unexported var and reused here.  According to the docs it is safe for concurrent use as well https://golang.org/pkg/regexp/#Regexp

File: agent/consul/catalog_endpoint.go [hunk 2]
```diff
@@ -97,6 +99,22 @@ func (c *Catalog) Register(args *structs.RegisterRequest, reply *struct{}) error
...
+		// Warn if any tags are incompatible with DNS
+		for _, tag := range args.Service.Tags {
+			if !dnsNameRe.MatchString(tag) {
+				return fmt.Errorf("Service tag contains invalid character for dns: %s for service %s,"+
+					"Blocking registration", tag, args.Service.Service)
+
+			}
+		}
+	}
+
```
Comment (preetapan) [inline agent/consul/catalog_endpoint.go]: Can you make a unit test for this?

**Comment by preetapan** [comment]
Also, could you explain if this is actually hurting anything in your environment vs you wanting a way to enforce the dns naming spec.

**Comment by magiconair** [comment]
When enabled this would prevent you from using https://github.com/fabiolb/fabio for example - and other services which put metadata in tags which isn't uncommon.

**Comment by vamage** [comment]
For our environment we use SRV records for managing a number of services(ie ngnix plus,puppet), Some services were added that didn't match is pattern and we wanted to ensure that services are all discoverable via DNS.

File: agent/consul/catalog_endpoint.go (hunk 2 above)
Comment (vamage) [inline agent/consul/catalog_endpoint.go]: Will do!

Comment (vamage) [inline agent/consul/catalog_endpoint.go]: Added test for both tags and services.

**Comment by i0rek** [comment]
Hello @vamage! Thanks for your contribution. I tried to rebase on current master, but it got out of hand. Would you be willing to do this again on current master, or rebase it by yourself or would you be ok if I would make the changes by myself?

Thanks again!

**Comment by i0rek** [comment]
Closing because no reply. Would be happy to reopen if someone wants to get this going again!

**Pull request closed without merging**